File: leetcode/0225_implement_stack_using_queues.cpp
// Copyright (c) 2018 Jacopo Notarstefano

#include <cassert>
#include <queue>

using namespace std;

class MyStack {
 public:
    MyStack() {}

    void push(int x) {
        queue_.push(x);
        for (int i = 0; i < queue_.size() - 1; i++) {
            int tmp = queue_.front();
            queue_.pop();
            queue_.push(tmp);
        }
    }

    int pop() {
        int tmp = queue_.front();
        queue_.pop();
        return tmp;
    }

    int top() {
        return queue_.front();
    }

    bool empty() {
        return queue_.empty();
    }

 private:
    queue<int> queue_;
};

int main() {
    auto obj = MyStack();

    obj.push(1);
    obj.push(2);
    assert(2 == obj.top());
    assert(2 == obj.pop());
    assert(!obj.empty());
}

File: leetcode/0225_implement_stack_using_queues.py
# -*- coding: utf-8 -*-

from collections import deque


class MyStack:
    def __init__(self):
        self.queue = deque()

    def push(self, x):
        self.queue.appendleft(x)
        for _ in range(len(self.queue) - 1):
            self.queue.appendleft(self.queue.pop())

    def pop(self):
        return self.queue.pop()

    def top(self):
        return self.queue[-1]

    def empty(self):
        return not self.queue


if __name__ == '__main__':
    obj = MyStack()

    obj.push(1)
    obj.push(2)
    assert 2 == obj.top()
    assert 2 == obj.pop()
    assert not obj.empty()
